### PR TITLE
205 infinite loops and potential memory leaks when running calctempscalcgw

### PR DESCRIPTION
**Pull request by vollous**
## Description

- Fix temperature infinite loops.
- Give ID to patched phase.

## Type of Change

 - [ ] Break (major)
 - [ ] Feature (minor)
 - [x] Bug fix (patch)
 - [ ] No source changes (no version change)

### CODE REVIEW

File: src/prog/CalcGW.cpp [hunk 1]
```diff
@@ -469,15 +464,23 @@ CLIOptions::CLIOptions(const BSMPT::parser &argparser)
...
     {
-      UseMultiStepPTMode = std::stoi(multistepPT_string);
+      int intUseMultiStepPTMode = std::stoi(multistepPT_string);
+      switch (intUseMultiStepPTMode)
+      {
+      case 0: UseMultiStepPTMode = MultiStepPTMode::OneStep; break;
+      case 1: UseMultiStepPTMode = MultiStepPTMode::EdgeCoverage; break;
+      case 2: UseMultiStepPTMode = MultiStepPTMode::CompleteCoverage; break;
+
+      default: ss << "--multistepmode invalid using 'default' instead"; break;
```
Comment (phbasler) [inline src/prog/CalcGW.cpp]:
```suggestion
      default: 
         { 
            ss << "--multistepmode invalid using 'default' instead"; 
            UseMultiStepPTMode = MultiStepPTMode:Default;
            break;
           }
```

I would set it explicit here, just to be sure

File: src/prog/CalcTemps.cpp [hunk 2]
```diff
@@ -433,15 +428,23 @@ CLIOptions::CLIOptions(const BSMPT::parser &argparser)
...
     {
-      UseMultiStepPTMode = std::stoi(multistepPT_string);
+      int intUseMultiStepPTMode = std::stoi(multistepPT_string);
+      switch (intUseMultiStepPTMode)
+      {
+      case 0: UseMultiStepPTMode = MultiStepPTMode::OneStep; break;
+      case 1: UseMultiStepPTMode = MultiStepPTMode::EdgeCoverage; break;
+      case 2: UseMultiStepPTMode = MultiStepPTMode::CompleteCoverage; break;
+
+      default: ss << "--multistepmode invalid using default instead"; break;
```
Comment (phbasler) [inline src/prog/CalcTemps.cpp]: Here also set default

File: src/prog/MinimaTracer.cpp [hunk 3]
```diff
@@ -422,15 +417,23 @@ CLIOptions::CLIOptions(const BSMPT::parser &argparser)
...
     {
-      UseMultiStepPTMode = std::stoi(multistepPT_string);
+      int intUseMultiStepPTMode = std::stoi(multistepPT_string);
+      switch (intUseMultiStepPTMode)
+      {
+      case 0: UseMultiStepPTMode = MultiStepPTMode::OneStep; break;
+      case 1: UseMultiStepPTMode = MultiStepPTMode::EdgeCoverage; break;
+      case 2: UseMultiStepPTMode = MultiStepPTMode::CompleteCoverage; break;
+
+      default: ss << "--multistepmode invalid using default instead"; break;
```
Comment (phbasler) [inline src/prog/MinimaTracer.cpp]: Here also set default

**Review by phbasler — CHANGES_REQUESTED** [review]

**Comment by phbasler** [comment]
@vollous Please have a look a the missing coverage and at the numeric compare remarks from copilot.

File: src/minimum_tracer/minimum_tracer.cpp [hunk 4]
```diff
@@ -2561,120 +2574,122 @@
...
 
-    int EdgeOfPhaseatTemp = 0;
-    for (auto temp : tempList)
+  int EdgeOfPhaseatTemp = 0;
+  for (auto temp : tempList)
+  {
+    for (auto edge : edgesList)
     {
-      for (auto edge : edgesList)
+      if (edge.temp == temp)
```
[the remark this answers is not in the document]
Comment (vollous) [reply src/minimum_tracer/minimum_tracer.cpp]: This would breake the logic, we want exact equalities.

File: src/minimum_tracer/minimum_tracer.cpp [hunk 5]
```diff
@@ -2501,8 +2526,9 @@
     Phase phase(Temp, T_high, T_low, MinTracer);
     addPhase(phase);
     print(phase);
-
-    MultiStepPTTracer(phase.T_low, -1);
+    // phase found and we managed to go lower in temperature
+    if (phase.MinimumPhaseVector.size() > 1 and phase.T_low != Temp)
```
[the remark this answers is not in the document]
Comment (vollous) [reply src/minimum_tracer/minimum_tracer.cpp]: This prevents infinite loops

**Comment by vollous** [comment]
> @vollous Please have a look a the missing coverage and at the numeric compare remarks from copilot.

The copilot fixes are not relevant, in fact they would break some logic. The coverage I am not sure how to improve, those are fallback for edge cases.

**Comment by vollous** [comment]
Let's see if this is enough for those edge cases.

**Comment by vollous** [comment]
> > Let's see if this is enough for those edge cases.
> 
> You are missing a test to trigger the `crit_status = BSMPT::StatusCrit::CoincideSinglePoint` case in the tracer, which I think was the point of this PR? Simple test to check if this returns the enum you expect to:
> 
>     * Setup an Instance of `CoexPhases` with T_high == T_low.
> 
>     * Check the crit_status
> 
> 
> As this does not calculate anything this test is instant and you are secured that this returns the wanted enum

I've added this and it slightly improved coverage but it seems that some tests (that locally at least add additional coverage) and not improving it... Particularly this https://github.com/phbasler/BSMPT/blob/7430b8c7399006137309f3b56014a28ffe16b543/src/minimum_tracer/minimum_tracer.cpp#L2648-L2673

**Review by phbasler — APPROVED** [review]
